tests/conftest.py

```diff
@@ -72,37 +77,79 @@ def drop_users_and_objects(cursor):
 
 
 @pytest.fixture
-def tiny_spec(tmpdir):
-    # NOTE: if the test_password isn't provided here we end up changing our test_user's password
-    # for real in our test_main_run_mode_works test below
+def base_spec(cursor):
+    """ A spec with the existing state of the test database before anything has been done """
+    spec = dedent("""
+        postgres:
+            attributes:
+                - BYPASSRLS
+                - CREATEDB
+                - CREATEROLE
+                - REPLICATION
+            can_login: true
+            is_superuser: true
+            owns:
+                schemas:
+                    - information_schema
+                    - pg_catalog
+                    - public
+                tables:
+                    - information_schema.*
+                    - pg_catalog.*
+            privileges:
+                schemas:
+                    write:
+                        - information_schema
+                        - pg_catalog
+                        - public
+
+        test_user:
+            attributes:
+                - PASSWORD "test_password"
+            can_login: yes
+            is_superuser: yes
+        """)
+
+    # Postgres 10 introduces several new roles that we have to account for
+    cursor.execute("SELECT substring(version from 'PostgreSQL ([0-9.]*) ') FROM version()")
+    pg_version = cursor.fetchone()[0]
+    if pg_version.startswith('10.'):
+        spec += dedent("""
```

Comment:
Adding notes for the public record:

Discussed this with Conrad in person. We both agree that there are pros and cons to this approach. The major pro is that we want to have end-to-end tests, which means we'll need to be reading in from a YAML spec.yml file somewhere within our tests. The downside is that this is a bit clunky and could be more cumbersome over time.

With that said, we both agreed that abstracting away from this would be premature optimization at this point. Besides that, we'd still need some sort of end-to-end test that uses YAML regardless.

As a result, we'll leave this as is for the moment but will be aware that if we add support for other databases (e.g. Redshift, Snowflake, etc.) that we may want to start reconsidering our approach to testing a bit.